Currency conversion tweaks

Tahvohck opened the pull request:
- update gitignore
- break tests out into single file
- better documentation
- file organized

Tahvohck committed:
break tests into standalone file

Tahvohck committed:
Move debug run-as-main into test

Tahvohck committed:
reorganize file

Tahvohck committed:
improve isAvailable logic

Tahvohck committed:
Improve documentation

Tahvohck committed:
Check if already loaded before first attempt to update

Tahvohck committed:
Avoid redeclaring every time function is called

These regexes never change, so they can be module-level

Tahvohck committed:
Add utility function

This adds a way to query the cache for all valid codes, since npm:currency-codes knows some our API doesn't, and our API knows some non-standard codes.

Tahvohck committed:
update ignore file

Tahvohck committed:
unify not-loaded tests

Tahvohck committed:
remove coverage comments that aren't working

Tahvohck committed:
lots of extra tests to improve coverage.

Mampfinator commented on src/CurrencyConversion.ts:
```diff
@@ -1,39 +1,90 @@
+ * @returns true if cache is loaded
+ */
+export function isLoaded() {
+    return ccCache != null;
```
This isn't quite right, `ccCache` is never `null`, only `undefined` (unassigned). Simplifying to `!ccCache`/changing `null` to `undefined` is probably better, even if it currently works because they're both falsish in the end.

Tahvohck replied:
fixed in 559a2b82982ad5ab0b655029efc1a83154e3dca2
Also `!ccCache` is incorrect, I think you meant `!!ccCache`

Mampfinator commented on src/CurrencyConversion.ts:
```diff
@@ -1,39 +1,90 @@
  * Loads the currency conversion cache from disk if available. If not, loads it from the API.
  * If the cache is out of date, updates it.
  */
 export async function loadCCCache() {
```
Since we're tweaking things anyway, having a way to force a cache refresh is probably a good idea down the road. So adding a `force` boolean parameter (false by default, obviously) might not be a bad idea.

Tahvohck replied:
Handled in 2050875

I'm not actually sure that there's any need to do this in actual program use, but it's not exactly a problem to add it.

Tahvohck committed:
fix variable definition check

undefined compared to null is not correct, double not returns a proper true/false

Tahvohck committed:
break bulky test up into sub-steps

Tahvohck committed:
Mock time to cover out-of-date cache

Tahvohck committed:
Test coverage no longer needed

Tahvohck committed:
firm up test code to be explicit about state.

Tahvohck committed:
Add force-reload option

Tahvohck committed:
firm up test flow

Tahvohck committed:
remove unneeded logging in test

Tahvohck committed:
Remove conditional that was forcing test run

I had put this in while doing debugging and forgot to remove it afterwards.

Tahvohck committed:
break test into sub-steps

Tahvohck committed:
Combine tests into one setup + substeps

Tahvohck committed:
fix bug in test: confirm conversion available

Tahvohck committed:
Test truncation done correctly

Tahvohck committed:
Use static data for testing where reasonable
